torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -129,7 +128,9 @@ def __init__(
             OrderedDict()
         )  # device, rank, params
         self._param_rank: Dict[torch.Tensor, int] = {}
+        self._param_to_index: Dict[int, int] = {}
```

Comment:
(this is prior to this PR) curious, what if we directly use `Tensor`  instead of `id(param)` as the key, will that result in wrong mapping? If so, shall we add a comment to mention that here?

Reply:
I saw the comments below. Would I be correct if I assume the concern of using Tensor as the map key was that Tensor hash values depends Tensor implementation which might 1) depend on Tensor value 2) be more expensive then `id()`?

Reply:
hmm good question, I didn't benchmark that, maybe nicer to move to Tensor for consistency ? The context is that for state dicts in pytorch code id(param) is typically used, and I did the same. When trying to "cache" this it moved into a more generic place, and it does look inconsistent with the line just above, I agree

Reply:
> I didn't benchmark that, maybe nicer to move to Tensor for consistency ? 

I am not sure how Tensor implementation computes its hash code. cc @albanD @gchanan do you know whether it's OK to use Tensor as map key?

Reply:
It just returns the id haha: https://github.com/pytorch/pytorch/blob/7a178a8a523d4653a3a2fa10c573b71e7fab1b9a/torch/tensor.py#L599-L602